File: languages/cpp/01-foundations/strings/example/string-clean-and-tokenize.cpp
#include <cctype>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    string line;
    cout << "Enter a sentence: ";
    getline(cin, line);

    string cleaned;
    cleaned.reserve(line.size());
    for (char ch : line) {
        if (isalnum(static_cast<unsigned char>(ch))) {
            cleaned.push_back(static_cast<char>(tolower(static_cast<unsigned char>(ch))));
        } else {
            cleaned.push_back(' ');
        }
    }

    vector<string> words;
    string currentWord;
    for (char ch : cleaned) {
        if (ch == ' ') {
            if (!currentWord.empty()) {
                words.push_back(currentWord);
                currentWord.clear();
            }
        } else {
            currentWord.push_back(ch);
        }
    }
    if (!currentWord.empty()) {
        words.push_back(currentWord);
    }

    cout << "Normalized text: " << cleaned << '\n';
    cout << "Tokens (" << words.size() << "):\n";
    for (const string& word : words) {
        cout << "- " << word << '\n';
    }

    return 0;
}

File: languages/go/01-foundations/strings/example/string-clean-and-tokenize.go
// This helper example focuses on isolating string cleanup before the main example combines the steps.

// This extra example extends strings with text normalization and token extraction.

package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter a sentence: ")
	line, err := reader.ReadString('\n')
	if err != nil && len(line) == 0 {
		fmt.Println("Could not read the sentence.")
		return
	}
	line = strings.TrimRight(line, "\r\n")

	cleanedRunes := make([]rune, 0, len(line))
	for _, char := range line {
		if unicode.IsLetter(char) || unicode.IsDigit(char) {
			cleanedRunes = append(cleanedRunes, unicode.ToLower(char))
		} else {
			cleanedRunes = append(cleanedRunes, ' ')
		}
	}

	cleaned := string(cleanedRunes)
	words := strings.Fields(cleaned)

	fmt.Printf("Normalized text: %s\n", cleaned)
	fmt.Printf("Tokens (%d):\n", len(words))
	for _, word := range words {
		fmt.Printf("- %s\n", word)
	}
}
